prereise/gather/griddata/hifld/data_process/transmission.py

```diff
@@ -538,32 +540,61 @@ def create_buses(lines):
     return buses
 
 
-def create_transformers(bus):
+def create_transformers(
+    bus, lines, transformer_designs, transformer_number_assumptions=None
+):
     """Add transformers between buses within the same substation. The assumed topology
     is that the highest-voltage bus in each substation is connected via a tansformer to
     every other voltage.
 
-    :param pandas.DataFrame bus: columns 'sub_id' and 'baseKV'
+    :param pandas.DataFrame bus: columns 'sub_id' and 'baseKV'.
+    :param pandas.DataFrame lines: columns 'from_bus_id', 'to_bus_id', and 'rateA'.
+    :param pandas.DataFrame transformer_designs: representative transformer data. Index
+        is (low_kV, high_kV), columns are 'x', 'r', and 'MVA'. 'x' and 'r' values are
+        per-unit, 'MVA' is in megawatts.
+    :param pandas.Series transformer_number_assumptions: assumptions about the number
+        of transformers between any two buses. The index should be a MultiIndex of
+        (sub_id, low_kV, high_kV), and values are integer. If None, this parameter will
+        not be used.
     :return: (*pandas.DataFrame*) -- each row is one transformer, columns are
-        ["from_bus_id", "to_bus_id"].
+        ["from_bus_id", "to_bus_id", "x", "r", and "rateA"].
     """
-    bus_pairs = [
-        (b, volt_series.sort_values().index[i + 1])
+    bus_pair_data = [
+        {
+            "sub_id": sub,
+            "from_bus_id": from_id,
+            "to_bus_id": to_id,
+            "from_kV": volt_series.loc[from_id],
+            "to_kV": volt_series.loc[to_id],
+        }
         for sub, volt_series in bus.groupby("sub_id")["baseKV"]
-        for i, b in enumerate(volt_series.sort_values().index[:-1])
-        if len(volt_series) > 1
+        for from_id, to_id in zip(volt_series.index[:-1], volt_series.index[1:])
```

Comment:
So we no longer need to sort by `baseKV` first? The default behavior of `zip` will stop until it reaches the end of the shorter list, then why we would like to exclude the last element explicitly?

Reply:
Aha, I missed that your refactor included the sorting before the groupby. The previous version works already because within `create_buses` we create buses within substations in sorted voltage order, so the buses within each substation are already sorted when they get here, but I think it's clearer to maintain the (redundant) sorting for the sake of being more flexible with the inputs.

I agree that we don't need to explicitly shorten both, but in my mind it's a little clearer what's going on here if we do so, without having to remember how `zip` functions for iterables of mixed length.